tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+
+2. `gridSetRowHomogeneous grid True` makes every row have equal height.
+
+3. Here we define the `attach` helper function. It attaches given widget to
+   our `grid`. The argument order of this function helps to use it with
+   `(>>=)`.
+
+4. We attach `display` we created previously to the `grid`. It will occupy
+   the entire top row.
+
+5. Here we use combination of `mkBtn` helper and `attach` to quickly create
+   buttons and place them on the grid. I'll show `mkBtn` is a moment.
+
+6. Now the grid itself needs to be inserted into `window` to be visible.
+   This is done with help of above-mentioned `containerAdd` function.
+
+`mkBtn` is a helper for button creation, right now it's very simple:
+
+```haskell
+mkBtn :: String -> IO Button
+mkBtn label = do
+  btn <- buttonNew
+  set btn [ buttonLabel := label ]
+  return btn
+```
+
+We create new button, set its attributes (just label in our case) and return
+the button.
+
+![Our calculator looks like a real calculator](/tutorials/haskell/gui-application/calc-2.png)
+
+## Signals and events
+
+The application looks like a calculator but does not behave like one yet. To
+fix this, we need to learn about *signals* and *events*.
+
+*Signal* is a name for things that may happen on the form. Almost always
+signals are connected with user's actions. An example of signal is *focus* —
+the moment when a widget becomes active on the form.
+
+To execute some code on signal we use the `on` helper:
+
+```haskell
+on :: object -> Signal object callback -> callback -> IO (ConnectId object)
+```
+
+Here `object` is the widget of interest, `callback` is the action that we
+want to perform. The `on` function returns `ConnectId` parametrized over the
+`object` type (so we cannot mix up connection identifiers for different
+types of objects). This is the identifier of signal handler and its sole
+purpose to give you a way to disconnect a signal handler if you ever need
+it. You can use the `disconnect` from `System.Glib.Signals` to do that.
+
+Every signal dictates type that `callback` function will have. The following
+cases are most frequent:
+
+* Just `IO ()`, no information is given to the handler it is not expected to
+  return anything. Example of such signal is `showSignal`.
+
+* Handlers that are given arguments: `a -> IO Bool`. Example of such signal
+  is `focus` whose handlers have the type `DirectionType -> IO Bool`.
+  Another interesting thing here is returned value of the type `Bool`. This
+  is a convention in GTK+ allowing to disable default handling of some
+  signals. If we return `True`, default handling will be disabled, while
+  `False` will keep it active executing our handler *and* default handler as
+  well.
+
+There is one more way to get some information from within signal's handler.
+Some signals dictate that handler should live in special monad called
+`EventM` instead of plain `IO`. Signals that like their handlers to be in
+`EventM` are called events.
```

Comment:
@mrkkrp 

- "from within **a** signal's handler".
- "in **a** special monad".